File: composeApp/src/androidMain/kotlin/co/jarias/flexapp/ui/screens/bingo/card_setup/BingoCardSetupScreenViewModel.kt
package co.jarias.flexapp.ui.screens.bingo.card_setup

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import co.jarias.flexapp.data.local.CardSetupState
import co.jarias.flexapp.data.local.PreferencesManager
import co.jarias.flexapp.data.repository.BingoCardRepository
import co.jarias.flexapp.domain.BingoCard
import co.jarias.flexapp.domain.BingoCell
import co.jarias.flexapp.domain.usecase.GenerateRandomNumbersUseCase
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class BingoCardSetupScreenViewModel(
    private val bingoCardRepository: BingoCardRepository,
    private val preferencesManager: PreferencesManager? = null,
    private val gameId: Long = 0
) : ViewModel() {

    private val _state = MutableStateFlow(BingoCardSetupScreenState())
    val state: StateFlow<BingoCardSetupScreenState> = _state.asStateFlow()

    private val generateRandomNumbersUseCase = GenerateRandomNumbersUseCase()

    init {
        if (gameId != 0L) {
            loadGame(gameId)
        }
    }

    private fun validateColumn(col: Int): Boolean {
        val cardNumbers = _state.value.cardNumbers
        val seen = mutableSetOf<Int>()
        for (row in 0..4) {
            if (row == 2 && col == 2) continue
            val value = cardNumbers[row][col]
            if (value.isBlank()) return false
            val num = value.toIntOrNull() ?: return false
            val ranges = listOf(1..15, 16..30, 31..45, 46..60, 61..75)
            if (num !in ranges.getOrElse(col) { 1..15 }) return false
            if (!seen.add(num)) return false
        }
        return true
    }

    private fun loadCardSetupStateFromPreferences() {
        viewModelScope.launch {
            val savedState = preferencesManager?.getCardSetupState(gameId)
            savedState?.let {
                _state.value = _state.value.copy(
                    currentColumn = it.currentColumn,
                    cardNumbers = it.cardNumbers
                )
            }
        }
    }

    private fun saveCardSetupStateToPreferences() {
        viewModelScope.launch {
            preferencesManager?.setCardSetupState(
                gameId,
                CardSetupState(
                    currentColumn = _state.value.currentColumn,
                    cardNumbers = _state.value.cardNumbers
                )
            )
        }
    }

    private fun clearCardSetupStateInPreferences() {
        viewModelScope.launch {
            preferencesManager?.clearCardSetupState(gameId)
        }
    }

    fun loadGame(gameId: Long) {
        viewModelScope.launch {
            try {
                _state.value = _state.value.copy(isLoading = true, gameId = gameId)

                val existingCard = withContext(Dispatchers.IO) {
                    bingoCardRepository.getCardByGameId(gameId)
                }

                if (existingCard != null) {
                    val cardNumbers = existingCard.grid.map { row ->
                        row.map { cell ->
                            cell.number?.toString() ?: ""
                        }
                    }
                    _state.value = _state.value.copy(
                        cardNumbers = cardNumbers,
                        isLoading = false
                    )
                    findFirstIncompleteColumn()
                } else {
                    loadCardSetupStateFromPreferences()
                    _state.value = _state.value.copy(isLoading = false)
                }
            } catch (e: Exception) {
                _state.value = _state.value.copy(
                    isLoading = false,
                    errorMessage = "Failed to load card: ${e.message}"
                )
            }
        }
    }

    private fun findFirstIncompleteColumn() {
        for (col in 0..4) {
            if (!validateColumn(col)) {
                _state.value = _state.value.copy(currentColumn = col)
                return
            }
        }
    }

    fun onEvent(event: BingoCardSetupScreenEvents) {
        when (event) {
            is BingoCardSetupScreenEvents.OnCardNumberChanged -> updateCardNumber(event.row, event.col, event.value)
            is BingoCardSetupScreenEvents.OnNextColumn -> nextColumn()
            is BingoCardSetupScreenEvents.OnPreviousColumn -> previousColumn()
            is BingoCardSetupScreenEvents.OnSaveCard -> saveCard()
            is BingoCardSetupScreenEvents.OnRandomFill -> randomFillCurrentColumn()
        }
    }

    private fun randomFillCurrentColumn() {
        val currentColumn = _state.value.currentColumn
        val randomNumbers = generateRandomNumbersUseCase(currentColumn)
        val numberList = randomNumbers.toList()

        val newCardNumbers = _state.value.cardNumbers.mapIndexed { rowIdx, row ->
            if (rowIdx == 2 && currentColumn == 2) {
                row
            } else {
                row.mapIndexed { colIdx, value ->
                    if (colIdx == currentColumn && value.isBlank()) {
                        val numIdx = if (currentColumn == 2) {
                            if (rowIdx < 2) rowIdx else rowIdx - 1
                        } else rowIdx
                        numberList.getOrNull(numIdx)?.toString() ?: value
                    } else value
                }
            }
        }

        _state.value = _state.value.copy(
            cardNumbers = newCardNumbers,
            errorMessage = null
        )
        saveCardSetupStateToPreferences()
    }

    fun updateCardNumber(row: Int, col: Int, value: String) {
        val newCardNumbers = _state.value.cardNumbers.mapIndexed { idx, list ->
            if (idx == row) {
                list.mapIndexed { c, v -> if (c == col) value else v }
            } else list
        }
        _state.value = _state.value.copy(
            cardNumbers = newCardNumbers,
            errorMessage = null
        )
        saveCardSetupStateToPreferences()
    }

    private fun nextColumn() {
        if (!validateColumn(_state.value.currentColumn)) {
            _state.value = _state.value.copy(
                errorMessage = "Please fill all fields with valid, unique numbers in the correct range."
            )
            return
        }

        saveCardSetupStateToPreferences()

        if (_state.value.currentColumn < 4) {
            _state.value = _state.value.copy(
                currentColumn = _state.value.currentColumn + 1,
                errorMessage = null
            )
        } else {
            saveCard()
        }
    }

    private fun previousColumn() {
        saveCardSetupStateToPreferences()

        if (_state.value.currentColumn > 0) {
            _state.value = _state.value.copy(
                currentColumn = _state.value.currentColumn - 1,
                errorMessage = null
            )
        }
    }

    private fun saveCard() {
        val currentState = _state.value

        if (!validateColumn(currentState.currentColumn)) {
            _state.value = currentState.copy(
                errorMessage = "Please fill all fields with valid, unique numbers in the correct range."
            )
            return
        }

        for (col in 0..4) {
            if (!validateColumn(col)) {
                _state.value = currentState.copy(
                    currentColumn = col,
                    errorMessage = "Please fill all fields with valid, unique numbers."
                )
                return
            }
        }

        viewModelScope.launch {
            try {
                _state.value = currentState.copy(isSaving = true)

                val grid = getCardNumbers()
                val card = BingoCard(id = null, gameId = currentState.gameId, grid = grid)

                withContext(Dispatchers.IO) {
                    bingoCardRepository.insertCard(card)
                }

                clearCardSetupStateInPreferences()

                _state.value = currentState.copy(
                    isSaving = false,
                    cardSaved = true
                )
            } catch (e: Exception) {
                _state.value = currentState.copy(
                    isSaving = false,
                    errorMessage = "Failed to save card: ${e.message}"
                )
            }
        }
    }

    private fun getCardNumbers(): List<List<BingoCell>> {
        return List(5) { row ->
            List(5) { col ->
                if (row == 2 && col == 2) {
                    BingoCell(number = null, isFree = true)
                } else {
                    val num = _state.value.cardNumbers[row][col].toIntOrNull() ?: 0
                    BingoCell(number = num, isFree = false)
                }
            }
        }
    }
}

File: iosApp/iosApp/Screens/bingo/card_setup/BingoCardSetupScreenViewModel.swift
import Foundation
import shared

class BingoCardSetupScreenViewModel: ObservableObject {
    @Published var state: BingoCardSetupScreenState
    
    private let getBingoCardUseCase: GetBingoCardUseCase
    private let saveBingoCardUseCase: SaveBingoCardUseCase
    private let getCardSetupStateUseCase: GetCardSetupStateUseCase
    private let saveCardSetupStateUseCase: SaveCardSetupStateUseCase
    private let clearCardSetupStateUseCase: ClearCardSetupStateUseCase
    private let generateRandomNumbersUseCase = GenerateRandomNumbersUseCase()
    
    init(
        getBingoCardUseCase: GetBingoCardUseCase,
        saveBingoCardUseCase: SaveBingoCardUseCase,
        getCardSetupStateUseCase: GetCardSetupStateUseCase,
        saveCardSetupStateUseCase: SaveCardSetupStateUseCase,
        clearCardSetupStateUseCase: ClearCardSetupStateUseCase,
        gameId: Int64
    ) {
        self.getBingoCardUseCase = getBingoCardUseCase
        self.saveBingoCardUseCase = saveBingoCardUseCase
        self.getCardSetupStateUseCase = getCardSetupStateUseCase
        self.saveCardSetupStateUseCase = saveCardSetupStateUseCase
        self.clearCardSetupStateUseCase = clearCardSetupStateUseCase
        self.state = BingoCardSetupScreenState(gameId: gameId)
        loadGame(gameId: gameId)
    }
    
    func onEvent(_ event: BingoCardSetupScreenEvents) {
        switch event {
        case .onCardNumberChanged(let row, let col, let value):
            updateCardNumber(row: row, col: col, value: value)
        case .onNextColumn:
            nextColumn()
        case .onPreviousColumn:
            previousColumn()
        case .onSaveCard:
            saveCard()
        case .onRandomFill:
            randomFillCurrentColumn()
        }
    }
    
    private func loadGame(gameId: Int64) {
        state.isLoading = true
        getBingoCardUseCase.invoke(gameId: gameId) { [weak self] existingCard, error in
            DispatchQueue.main.async {
                guard let self = self else { return }
                if let card = existingCard {
                    let cardNumbers = card.grid.map { row in
                        row.map { cell in
                            cell.number?.stringValue ?? ""
                        }
                    }
                    self.state.cardNumbers = cardNumbers
                    self.state.isLoading = false
                    self.findFirstIncompleteColumn()
                } else {
                    self.loadCardSetupStateFromPreferences()
                    self.state.isLoading = false
                }
            }
        }
    }
    
    private func loadCardSetupStateFromPreferences() {
        getCardSetupStateUseCase.invoke(gameId: state.gameId) { [weak self] savedState, error in
            DispatchQueue.main.async {
                guard let self = self else { return }
                if let it = savedState {
                    self.state.currentColumn = Int(it.currentColumn)
                    self.state.cardNumbers = it.cardNumbers
                }
            }
        }
    }
    
    private func saveCardSetupStateToPreferences() {
        let cardSetupState = CardSetupState(currentColumn: Int32(state.currentColumn), cardNumbers: state.cardNumbers)
        saveCardSetupStateUseCase.invoke(gameId: state.gameId, state: cardSetupState) { error in }
    }
    
    private func clearCardSetupStateInPreferences() {
        clearCardSetupStateUseCase.invoke(gameId: state.gameId) { error in }
    }
    
    private func findFirstIncompleteColumn() {
        for col in 0..<5 {
            if !validateColumn(col: col) {
                state.currentColumn = col
                return
            }
        }
    }
    
    private func validateColumn(col: Int) -> Bool {
        let cardNumbers = state.cardNumbers
        var seen = Set<Int>()
        for row in 0..<5 {
            if row == 2 && col == 2 { continue }
            let value = cardNumbers[row][col]
            if value.trimmingCharacters(in: .whitespaces).isEmpty { return false }
            guard let num = Int(value) else { return false }
            let ranges = [1...15, 16...30, 31...45, 46...60, 61...75]
            let range = ranges[col]
            if !range.contains(num) { return false }
            if seen.contains(num) { return false }
            seen.insert(num)
        }
        return true
    }
    
    private func updateCardNumber(row: Int, col: Int, value: String) {
        state.cardNumbers[row][col] = value
        state.errorMessage = nil
        saveCardSetupStateToPreferences()
    }
    
    private func nextColumn() {
        if !validateColumn(col: state.currentColumn) {
            state.errorMessage = "Please fill all fields with valid, unique numbers in the correct range."
            return
        }
        
        saveCardSetupStateToPreferences()
        
        if state.currentColumn < 4 {
            state.currentColumn += 1
            state.errorMessage = nil
        } else {
            saveCard()
        }
    }
    
    private func previousColumn() {
        saveCardSetupStateToPreferences()
        if state.currentColumn > 0 {
            state.currentColumn -= 1
            state.errorMessage = nil
        }
    }
    
    private func randomFillCurrentColumn() {
        let currentColumn = state.currentColumn
        let randomNumbers = generateRandomNumbersUseCase.invoke(columnIndex: Int32(currentColumn))
        let numberList = Array(randomNumbers).map { ($0 as? Int) ?? 0 }
        
        for rowIdx in 0..<5 {
            if rowIdx == 2 && currentColumn == 2 { continue }
            if state.cardNumbers[rowIdx][currentColumn].isEmpty {
                let numIdx: Int
                if currentColumn == 2 {
                    numIdx = rowIdx < 2 ? rowIdx : rowIdx - 1
                } else {
                    numIdx = rowIdx
                }
                if numIdx < numberList.count {
                    state.cardNumbers[rowIdx][currentColumn] = String(numberList[numIdx])
                }
            }
        }
        
        state.errorMessage = nil
        saveCardSetupStateToPreferences()
    }
    
    private func saveCard() {
        if !validateColumn(col: state.currentColumn) {
            state.errorMessage = "Please fill all fields with valid, unique numbers in the correct range."
            return
        }
        
        for col in 0..<5 {
            if !validateColumn(col: col) {
                state.currentColumn = col
                state.errorMessage = "Please fill all fields with valid, unique numbers."
                return
            }
        }
        
        state.isSaving = true
        let grid = getBingoGrid()
        let card = BingoCard(id: nil, gameId: state.gameId, grid: grid)
        
        saveBingoCardUseCase.invoke(card: card) { [weak self] error in
            DispatchQueue.main.async {
                guard let self = self else { return }
                self.state.isSaving = false
                if let error = error {
                    self.state.errorMessage = "Failed to save card: \(error.localizedDescription)"
                } else {
                    self.clearCardSetupStateInPreferences()
                    self.state.cardSaved = true
                }
            }
        }
    }
    
    private func getBingoGrid() -> [[BingoCell]] {
        return (0..<5).map { row in
            (0..<5).map { col in
                if row == 2 && col == 2 {
                    return BingoCell(number: nil, isMarked: false, isFree: true)
                } else {
                    let num = Int32(state.cardNumbers[row][col]) ?? 0
                    return BingoCell(number: KotlinInt(value: num), isMarked: false, isFree: false)
                }
            }
        }
    }
}
